clippy_lints/src/misc.rs

```diff
@@ -509,7 +509,7 @@ fn is_allowed<'a, 'tcx>(cx: &LateContext<'a, 'tcx>, expr: &'tcx Expr<'_>) -> boo
             Constant::F64(f) => *f == 0.0 || (*f).is_infinite(),
             _ => false,
         }),
-        _ => false,
+        Some(_) | None => false,
```

Comment:
This shouldn't lint here, since `Some(_)` and `None` is left.

Reply:
Current implementation only checks if a variant exists and doesn't check if it's exsaustive. I guess it was ok for wildcard_enum_match_arm, but it's not ok in case of the current lint. I'll try to implement it.
I don't know if there can be cases when it's not possible to check for exuastiveness, but if there is, what should I do then? Should I just skip that match?

I'm sorry, I guess I should have opened a WIP PR

Reply:
We could also implement the same logic as with skipping arm guards. 
If a variant is appearing only in patterns with arm guards, then it is treated as missing.
https://github.com/rust-lang/rust-clippy/blob/master/clippy_lints/src/matches.rs#L722
```
            if arm.guard.is_some() {
                // Guards mean that this case probably isn't exhaustively covered. Technically
                // this is incorrect, as we should really check whether each variant is exhaustively
                // covered by the set of guards that cover it, but that's really hard to do.
                eprintln!("Arm: {:#?}", arm);
                continue;
            }
```
We could check only for simple case Variant(wild), and if it doesn't match then don't count Variant as present.

Which way do you want to go?

Reply:
> I'm sorry, I guess I should have opened a WIP PR

Not a problem, doesn't really matter review wise.

> Which way do you want to go?

The way you think is the easiest way to implement this. I see you put more thought into this, than I have, so I trust your judgement here.

In the end it should only suggest to replace it with the remaining variant, if we can be absolutely sure that there is only one left.

Reply:
I've fixed it. Currently variants with TupleStruct patterns are treated as missing unless they pass checks for exhaustiveness. The checks are simple. They check for wild and binding patterns.